### ISSUE TITLE
jsonencode() should support a single quoted value (this is valid JSON)

### ISSUE THREAD

**Issue by voycey** [hashicorp/terraform#32032]
### Terraform Version

```shell
Terraform v1.2.9
on linux_amd64
+ provider registry.terraform.io/hashicorp/github v5.3.0
+ provider registry.terraform.io/hashicorp/google v3.90.1
+ provider registry.terraform.io/hashicorp/google-beta v3.90.1
+ provider registry.terraform.io/hashicorp/kubernetes v2.14.0
+ provider registry.terraform.io/hashicorp/random v3.4.3
+ provider registry.terraform.io/hashicorp/vault v3.9.1
+ provider registry.terraform.io/newrelic/newrelic v3.4.4

Your version of Terraform is out of date! The latest version
is 1.3.2. You can update by downloading from https://www.terraform.io/downloads.html
```

### Use Cases

Here is an example of where this would come in very handy:

```terraform
resource "vault_kv_secret_v2" "non-secrets" {
  mount = vault_mount.kvv2.path
  for_each  = var.variables
  name = "${var.app_name}-${var.stage}/${each.key}"
  cas = 1
  delete_all_versions = true

  data_json = jsonencode(each.value)
}
```

when writing out KV pairs - quite often the value should just be a string, not necessarily some other higher form of JSON object.

This has been discussed at length online: https://stackoverflow.com/a/22361229/2748998 and various other places, often coming down to the difference being "a JSON object" vs "Valid JSON" 

`"I am valid JSON"`

### Attempted Solutions

```terraform
resource "vault_kv_secret_v2" "non-secrets" {
  mount = vault_mount.kvv2.path
  for_each  = var.variables
  name = "${var.app_name}-${var.stage}/${each.key}"
  cas = 1
  delete_all_versions = true

  data_json = each.value
}
```

```terraform
resource "vault_kv_secret_v2" "non-secrets" {
  mount = vault_mount.kvv2.path
  for_each  = var.variables
  name = "${var.app_name}-${var.stage}/${each.key}"
  cas = 1
  delete_all_versions = true

  data_json = "${each.value}"
}
```

And various other types of ways to get a single string in there.

This usually ends up with an unmarshalling errror:

`json: cannot unmarshal string into Go value of type map[string]interface {}`

or

```
 Error: Invalid function argument
│
│   on modules/vault-output/vault-output.tf line 24, in resource "vault_kv_secret_v2" "non-secrets":
│   24:   data_json = jsonencode(tostring(each.value))
│     ├────────────────
│     │ each.value is object with 3 attributes
│
│ Invalid value for "v" parameter: cannot convert object to string.
```

The following works:

```terraform
resource "vault_kv_secret_v2" "non-secrets" {
  mount = vault_mount.kvv2.path
  for_each  = var.variables
  name = "${var.app_name}-${var.stage}/${each.key}"
  cas = 1
  delete_all_versions = true

  data_json = jsonencode({
    "${each.key}" = each.value
  })
}
```

But this would mean accessing all variables via `keyname.keyname`

### Proposal

Simple enable `jsonencode(string)` and allow that to be returned as a JSON string representation 

e.g

```
each.value = "apm_config"
jsonencode(each.value)

"apm_config"
```

This would solve any number of use cases

**Comment by apparentlymart** [comment]
Hi @voycey! Thanks for reporting this.

I believe the behavior you've seen here belongs to the provider you are using, rather than to Terraform's `jsonencode` function.

Notice that the error message you shared talks about _unmarshalling_ (decoding) a string, not about encoding one. This suggests that the `jsonencode` function did successfully produce a JSON-encoded string, but that the provider then tried to decode it into a map (apparently using the Go `encoding/json` library, and so the error message confusingly talks about Go types instead of Terraform types.)

The `jsonencode` function already accepts values of any type, so I think this is a request for the `hashicorp/vault` provider instead of Terraform Core. Therefore I'm going to close this issue. I suggest opening a similar request in the provider's own GitHub repository, which you can find by looking up the provider in [Terraform Registry](https://registry.terraform.io/).

Thanks!

**Comment by voycey** [comment]
Thanks @apparentlymart! Yep this is totally on the provider - the error messages get pretty hard to follow this deep down the rabbit hole 😅
